### DISCUSSION TITLE
What is the right way to have GitVersion report versioning from a different branch pointer?

### DISCUSSION THREAD

**Discussion by hotchkj** [GitTools/GitVersion#4160] [Q&A] [answered]
I'm implementing some more advanced versioning reasoning in a build for pull requests. I want to know the version in the PR, _and_ the version in the PR target branch.

Example: in Jenkins, I can have `CHANGE_TARGET=master`, while the actual code is checked out on the PR merge branch `PR-1234`. I'd like GitVersion to calculate version for the `master` HEAD as well as the `PR-1234` HEAD.

What is the best/right way to do this? Are there missing features that would make this easier? I came across #3954 which suggests that I can't just temporarily override `GIT_BRANCH`.

I could probably just check out said branch and re-run GitVersion, but that's going to yield a _lot_ of expensive churn in the repo. I am curious if there's a less brute-force approach.

Thanks for assistance!

**Comment by asbjornu** [comment, marked as answer]
We are discussing similar use of `Git_Branch` in https://github.com/GitTools/GitVersion/issues/4212#issuecomment-2531239602. I believe that if implemented, it would cover your use-case as well.

**Comment by HHobeck** [comment]
> I'm implementing some more advanced versioning reasoning in a build for pull requests. I want to know the version in the PR, and the version in the PR target branch.

What do you mean you want to know the version in the PR target branch!? Can you give an example? What is the motivation behind?

**Reply by hotchkj** [reply]
Exactly as written. A pull request is by definition version A merging into version B. I want to be able to calculate both, for example for API compatibility checks.

From a Git perspective, this is examining the version history from the perspective of two different branches, `PR-1234` or `target`, usually `master`. A calculation of latest PR might be `v1.2.3-pullrequest0003`, merging into what is currently `v1.2.2`.

As the above alludes to, if `GIT_BRANCH` were authoritative, the GitVersion algorithm could just be re-run with different `GIT_BRANCH` settings. That would seem to do it.

Another way to think about this as a feature is 'I want to run GitVersion and specify the branch used to calculate the resulting version rather than have GitVersion guess that for me'.

**Reply by HHobeck** [reply]
Thank you for clarification. That means the solution for your scenario is to use the Git_Branch environment variable right? For me it would make no sense to calculate it twice by default. The only thing which I can think of is to have a output variable which indicates whether it is a none, patch, minor or major increment.

May I ask you to resolve or to close this discussion?

Have a nice day.
